src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+                                         circpad_delay_t target_bin_us);
+void circpad_machine_remove_closest_token(circpad_machineinfo_t *mi,
+                                          circpad_delay_t target_bin_us,
+                                          int use_usec);
+STATIC void circpad_machine_setup_tokens(circpad_machineinfo_t *mi);
+static inline circpad_purpose_mask_t circpad_circ_purpose_to_mask(uint8_t
+                                          circ_purpose);
+static inline circpad_circuit_state_t circpad_circuit_state(
+                                        origin_circuit_t *circ);
+static void circpad_circuit_machineinfo_free_idx(circuit_t *circ, int idx);
+static void circpad_setup_machine_on_circ(circuit_t *on_circ,
+                                          circpad_machine_t *machine);
+static double circpad_distribution_sample(circpad_distribution_t dist);
+
+/** Cached consensus params */
+static uint8_t circpad_global_max_padding_percent;
+static uint16_t circpad_global_allowed_cells;
+
+/** Global cell counts, for rate limiting */
+static uint64_t circpad_global_padding_sent;
+static uint64_t circpad_global_nonpadding_sent;
+
+/** This is the list of circpad_machine_t's parsed from consensus and torrc
+ *  that have origin_side == 1 (ie: are for client side) */
+static smartlist_t *origin_padding_machines = NULL;
+
+/** This is the list of circpad_machine_t's parsed from consensus and torrc
+ *  that have origin_side == 0 (ie: are for relay side) */
+static smartlist_t *relay_padding_machines = NULL;
+
+/** Loop over the current padding state machines using <b>loop_var</b> as the
+ *  loop variable. */
+#define FOR_EACH_CIRCUIT_MACHINE_BEGIN(loop_var)                         \
+  STMT_BEGIN                                                             \
+  for (int loop_var = 0; loop_var < CIRCPAD_MAX_MACHINES; loop_var++) {
+#define FOR_EACH_CIRCUIT_MACHINE_END } STMT_END ;
+
+/** Loop over the current active padding state machines using <b>loop_var</b>
+ *  as the loop variable. If a machine is not active, skip it. */
+#define FOR_EACH_ACTIVE_CIRCUIT_MACHINE_BEGIN(loop_var, circ)            \
+  FOR_EACH_CIRCUIT_MACHINE_BEGIN(loop_var)                               \
+  if (!(circ)->padding_info[loop_var])                           \
+    continue;
+#define FOR_EACH_ACTIVE_CIRCUIT_MACHINE_END } STMT_END ;
+
+/**
+ * Return a human-readable description for a circuit padding state.
+ */
+static const char *
+circpad_state_to_string(circpad_statenum_t state)
+{
+  const char *descr;
+
+  switch (state) {
+  case CIRCPAD_STATE_START:
+    descr = "START";
+    break;
+  case CIRCPAD_STATE_BURST:
+    descr = "BURST";
+    break;
+  case CIRCPAD_STATE_GAP:
+    descr = "GAP";
+    break;
+  case CIRCPAD_STATE_END:
+    descr = "END";
+    break;
+  default:
+    descr ="INVALID";
+  }
+
+  return descr;
+}
+
+/**
+ * Return the circpad_stat_t for the current state based on the
```

Comment:
stat -> state.

Also let's document that it can return NULL, and when it does so.

Reply:
Addressed in 01ef88908a.

Reply:
Addressed in 3ee2f69ddb5c9a02c4d15a6908ea4cf53eaf7c93.